src/org/zaproxy/zap/extension/ascanrulesBeta/InsecureHTTPMethod.java

```diff
@@ -1,455 +1,526 @@
+
+
+                // use a CONNECT method to try establish a socket connection to
+                // a third party, via the server being tested
+                testConnect(this.getBaseMsg(), thirdpartyHost, thirdpartyPort, thirdPartyContentPattern);
+
+                /* Test all other methods */
+                for (String insecureMethod : INSECURE_METHODS) {
+                    testHttpMethod(insecureMethod);
+                }
+            } else { // Not HIGH or INSANE
+                // send an OPTIONS message, and see what the server reports. Do
+                // not try any methods not listed in those results.
+                HttpMessage optionsmsg = getNewMsg();
+                HttpRequestHeader optionsRequestHeader = this.getBaseMsg().getRequestHeader();
+                optionsRequestHeader.setMethod(HttpRequestHeader.OPTIONS);
+                // OPTIONS is not supported in 1.0
+                optionsRequestHeader.setVersion(HttpRequestHeader.HTTP11);
+                optionsmsg.setRequestHeader(optionsRequestHeader);
+
+                sendAndReceive(optionsmsg, false); // do not follow redirects
+
+                allowedmethods = optionsmsg.getResponseHeader().getHeader(HttpHeader.METHODS_ALLOW);
+                publicmethods = optionsmsg.getResponseHeader().getHeader(HttpHeader.METHODS_PUBLIC);
+
+                /*
+                 * //DEBUG only, to test the CONNECT method against a Squid
+                 * instance, which does not support OPTIONS. //TODO: need to
+                 * test for these insecure methods, even if the OPTIONS method
+                 * did not indicate that the method is supported //ie, test for
+                 * hidden/masked support for these insecure methods, as well as
+                 * documented support.
+                 * log.error("Setting the allowed methods to 'CONNECT'");
+                 * allowedmethods = "CONNECT"; publicmethods = null;
+                 */
+
+                if (log.isDebugEnabled()) {
+                    log.debug("allowedmethods: " + allowedmethods);
+                    log.debug("publicmethods: " + publicmethods);
+                }
+
+                if (allowedmethods == null) {
+                    // nothing to see here. Move along now.
+                    return;
+                }
+                // if the "Public" response is present (for IIS), use that to
+                // determine the enabled methods.
+                if (publicmethods != null) {
+                    allowedmethods = publicmethods;
+                }
+
+                // Convert the list to a set so that we have a unique list
+                Set<String> enabledMethodsSet = new HashSet<>(Arrays.asList(allowedmethods.toUpperCase().split(",")));
+                // rely on the OPTIONS METHOD, but potentially verify the
+                // results, depending on the Threshold.
+                for (String enabledmethod : enabledMethodsSet) {
```

Comment:
I changed this to use a set in case there were duplicate entries.

I encountered `Allow: POST,OPTIONS,HEAD,HEAD,GET,HEAD,TRACE` while testing against a local XAMPP.